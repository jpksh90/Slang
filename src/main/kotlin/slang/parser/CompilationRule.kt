package slang.parser

import SlangBaseListener
import org.antlr.v4.runtime.ParserRuleContext
import org.antlr.v4.runtime.tree.ParseTreeWalker
import org.slf4j.LoggerFactory

/**
 * Annotation to define a parser rule.
 *
 * @property name The name of the parser rule.
 * @property description A brief description of the parser rule.
 * @property enabled Indicates if the parser rule is enabled. Default is true.
 */
@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
annotation class ParserRule(val name: String, val description: String, val enabled: Boolean = true)

/**
 * Base class for compilation rules.
 *
 * @property errorListener The error listener to log compilation errors.
 */
abstract class CompilationRule(private val errorListener: SlangParserErrorListener) : SlangBaseListener() {
    private var result: Boolean = true

    /**
     * Applies the compilation rule to the given parser context.
     *
     * @param ctx The parser context to apply the rule to.
     * @return True if the rule passes, false otherwise.
     */
    fun apply(ctx: ParserRuleContext): Boolean {
        ParseTreeWalker().walk(this, ctx)
        return result
    }

    /**
     * Logs a compilation error.
     *
     * @param error The error to log.
     */
    private fun logCompilationError(error: Error) {
        result = false
        errorListener.addError(error)
    }

    /**
     * Logs a compilation error with specific details.
     *
     * @param line The line number where the error occurred.
     * @param charPositionInLine The character position in the line where the error occurred.
     * @param message The error message.
     */
    fun logCompilationError(lineColumn: LineColumn, message: String) {
        logCompilationError(Error(lineColumn, message))
    }
}


/**
 * Manages the compilation rules.
 */
class CompilationRuleManager {
    private val rules = mutableListOf<CompilationRule>()

    private val logger = LoggerFactory.getLogger(CompilationRule::class.java)


    /**
     * Adds a compilation rule to the manager.
     *
     * @param rule The compilation rule to add.
     */
    fun addRule(rule: CompilationRule) {
        rules.add(rule)
    }

    /**
     * Applies all the compilation rules to the given parser context.
     *
     * @param ctx The parser context to apply the rules to.
     * @return True if all rules pass, false otherwise.
     */
    fun applyRules(ctx: ParserRuleContext): Boolean {
        return rules.fold(true) { acc, rule ->
            val annotation = rule::class.java.getAnnotation(ParserRule::class.java)
            if (annotation != null) {
                logger.info("Applying rule: ${annotation.name}")
            }
            acc && rule.apply(ctx)
        }
    }
}
