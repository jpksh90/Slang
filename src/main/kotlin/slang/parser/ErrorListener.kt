package slang.parser

import org.antlr.v4.runtime.BaseErrorListener
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.Recognizer


data class Error(val line: Int, val charPositionInLine: Int, val message: String) {
    override fun toString(): String {
        return "$line:$charPositionInLine - $message"
    }
}

class SlangParserErrorListener : BaseErrorListener() {
    val errors = mutableListOf<Error>()

    override fun syntaxError(
        recognizer: Recognizer<*, *>?,
        offendingSymbol: Any?,
        line: Int,
        charPositionInLine: Int,
        msg: String?,
        e: RecognitionException?
    ) {
        val error = Error(line, charPositionInLine, msg ?: "Unknown error")
        errors.add(error)
    }

    fun addError(line: Int, charPositionInLine: Int, message: String) {
        errors.add(Error(line, charPositionInLine, message))
    }
}