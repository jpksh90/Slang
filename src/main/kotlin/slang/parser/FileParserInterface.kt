package slang.parser

import SlangLexer
import SlangParser
import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CharStream
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.ParserRuleContext
import org.reflections.Reflections
import org.slf4j.LoggerFactory
import java.io.File

data class LineColumn(val line: Int, val column: Int) : Comparable<LineColumn> {
    override fun compareTo(other: LineColumn): Int {
        return if (line == other.line) {
            column - other.column
        } else {
            line - other.line
        }
    }

    override fun toString(): String {
        return "$line:$column"
    }
}

fun ParserRuleContext.lineColumn() = LineColumn(this.start.line, this.start.charPositionInLine)

abstract class Parser(input: CharStream) {

    var compilationUnit: SlangParser.CompilationUnitContext
    var parser: SlangParser
    private val errorListener = SlangParserErrorListener()
    private val ruleManager = CompilationRuleManager()

    private val logger = LoggerFactory.getLogger(FileParserInterface::class.java)

    init {
        val lexer = SlangLexer(input)
        parser = SlangParser(CommonTokenStream(lexer))
        parser.addErrorListener(SlangParserErrorListener())
        parser.addErrorListener(errorListener)
        compilationUnit = parser.compilationUnit()
        initializeCompilationRules()
    }

    private fun initializeCompilationRules() {
        val reflections = Reflections("slang.parser")
        for (rule in reflections.getSubTypesOf(CompilationRule::class.java)) {
            val annotation = rule.getAnnotation(ParserRule::class.java) ?: continue

            if (!annotation.enabled) {
                logger.warn("Rule ${rule.simpleName} is not enabled.")
                continue
            }

            val constructor = rule.getConstructor(SlangParserErrorListener::class.java)
            val ruleInstance = constructor.newInstance(errorListener)
            ruleManager.addRule(ruleInstance)
        }
    }

    // This method is used to parse the compilation unit and apply the rules. Basic syntax has been handled by Antlr
    fun parse() : Boolean {
        ruleManager.applyRules(compilationUnit)
        return errorListener.errors.isEmpty()
    }

    fun getErrors(): List<Error> {
        return errorListener.getErrorsSorted()
    }
}

class FileParserInterface(file: File) : Parser(ANTLRInputStream(file.inputStream().bufferedReader()))

class StringParserInterface(program: String) : Parser(ANTLRInputStream(program))

class URIParserInterface(uri: String) : Parser(ANTLRInputStream(uri.byteInputStream().bufferedReader()))
