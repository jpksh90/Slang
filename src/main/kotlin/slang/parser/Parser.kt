package slang.parser

import SlangBaseVisitor
import SlangLexer
import SlangParser
import SlangParser.ExprContext
import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CommonTokenStream
import java.io.File

abstract class CompilationRule(val errorListener: SlangParserErrorListener) : SlangBaseVisitor<Boolean>()

class Parser(file: File) {
    var compilationUnit: SlangParser.CompilationUnitContext? = null
    var parser: SlangParser
    val errorListener = SlangParserErrorListener()

    val compilationRules = mutableListOf<CompilationRule>()

    init {
        val input = file.inputStream().bufferedReader().use { it.readText() }
        val lexer = SlangLexer(ANTLRInputStream(input))
        parser = SlangParser(CommonTokenStream(lexer))
        parser.addErrorListener(SlangParserErrorListener())
        parser.addErrorListener(errorListener)
        compilationUnit = parser.compilationUnit()

        // Initialize the compilation rules here
        addCompilationRule(BreakContinueChecker(errorListener))
        addCompilationRule(AnonymousFunctionInBinaryOperator(errorListener))
    }

    fun getErrors() : List<String> {
        return errorListener.errors
    }

    private fun addCompilationRule(rule: CompilationRule) {
        compilationRules.addLast(rule)
    }

    fun parse(): SlangParser.CompilationUnitContext? {
        // evaluation of all compilation rules on the compilation unit should be true
        val compilationRulesStatus = compilationRules.fold(true) { acc, rule ->
            acc && rule.visit(compilationUnit)
        }

        if (compilationRulesStatus == false) {
            compilationUnit = null
        }

        return compilationUnit
    }
}

class BreakContinueChecker(errorListener: SlangParserErrorListener) : CompilationRule(errorListener) {
    private val loopContextStack = ArrayDeque<Boolean>()

    override fun visitWhileStmt(ctx: SlangParser.WhileStmtContext): Boolean {
        loopContextStack.add(true)
        val result = visitChildren(ctx)
        loopContextStack.removeAt(loopContextStack.size - 1)
        return result
    }

    override fun visitForStmt(ctx: SlangParser.ForStmtContext): Boolean {
        loopContextStack.add(true)
        val result = visitChildren(ctx)
        loopContextStack.removeAt(loopContextStack.size - 1)
        return result
    }

    override fun visitDoWhileStmt(ctx: SlangParser.DoWhileStmtContext?): Boolean {
        loopContextStack.add(true)
        val result = visitChildren(ctx)
        loopContextStack.removeAt(loopContextStack.size - 1)
        return result
    }

    override fun visitBreakStmt(ctx: SlangParser.BreakStmtContext): Boolean {
        if (loopContextStack.isEmpty()) {
            errorListener.errors.add("Break statement not within a loop at line ${ctx.start.line}:${ctx.start
                .charPositionInLine} - ${ctx.text}")
        }
        return visitChildren(ctx)
    }

    override fun visitContinueStmt(ctx: SlangParser.ContinueStmtContext): Boolean {
        if (loopContextStack.isEmpty()) {
            errorListener.errors.add("Continue statement not within a loop at line ${ctx.start.line}:${ctx.start
                .charPositionInLine} - ${ctx.text}")
        }
        return visitChildren(ctx)
    }
}

class AnonymousFunctionInBinaryOperator(errorListener: SlangParserErrorListener) : CompilationRule(errorListener) {

    private fun isAnonymousFunction(ctx: SlangParser.ExprContext) : Boolean {
        return ctx is SlangParser.FunAnonymousPureExprContext || ctx is SlangParser.FunAnonymousImpureExprContext
    }

    private fun check(operand1: SlangParser.ExprContext, operand2: ExprContext, type: String) : Boolean {
        if (isAnonymousFunction(operand1)) {
            errorListener.errors.add(
                "Anonymous function cannot be used as operand in $type expression at line " +
                        "${operand1.start.line}:${operand1.start.charPositionInLine} - ${operand1.text}"
            )
            return false
        }
        if (isAnonymousFunction(operand2)) {
            errorListener.errors.add(
                "Anonymous function cannot be used as operand in $type expression at line " +
                        "${operand2.start.line}:${operand2.start.charPositionInLine} - ${operand2.text}"
            )
            return false
        }
        return true
    }


    override fun visitArithmeticExpr(ctx: SlangParser.ArithmeticExprContext): Boolean {
        if (ctx.expr().size != 2) {
            check(ctx.expr(0), ctx.expr(1), "arithmetic")
            return visitChildren(ctx)
        }
        return visitChildren(ctx)
    }

    override fun visitComparisonExpr(ctx: SlangParser.ComparisonExprContext): Boolean {
        if (ctx.expr().size == 2) {
            check(ctx.expr(0), ctx.expr(1), "comparison")
            return visitChildren(ctx)
        }
        return true; // it's ok to pass the rules
    }

    override fun visitBooleanExpr(ctx: SlangParser.BooleanExprContext): Boolean {
        if (ctx.expr().size != 2) {
            check(ctx.expr(0), ctx.expr(1), "boolean")
            return visitChildren(ctx)
        }
        return visitChildren(ctx)
    }
}