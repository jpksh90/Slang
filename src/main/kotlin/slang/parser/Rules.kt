package slang.parser

import SlangParser
import org.antlr.v4.runtime.ParserRuleContext

@ParserRule("DisallowedBreakContinue", "Checks for break and continue statements outside of loops")
class BreakContinueChecker(errorListener: SlangParserErrorListener) : CompilationRule(errorListener) {
    private val loopContextStack = ArrayDeque<Boolean>()

    override fun enterWhileStmt(ctx: SlangParser.WhileStmtContext) {
        loopContextStack.add(true)
    }

    override fun exitWhileStmt(ctx: SlangParser.WhileStmtContext) {
        loopContextStack.removeLast()
    }

    override fun enterDoWhileStmt(ctx: SlangParser.DoWhileStmtContext) {
        loopContextStack.add(true)
    }

    override fun exitDoWhileStmt(ctx: SlangParser.DoWhileStmtContext) {
        loopContextStack.removeLast()
    }

    override fun enterBreakStmt(ctx: SlangParser.BreakStmtContext) {
        if (loopContextStack.isEmpty()) {
            logCompilationError(ctx.toLineColumn(), "Break statement not within a loop")
        }
    }

    override fun enterContinueStmt(ctx: SlangParser.ContinueStmtContext) {
        if (loopContextStack.isEmpty()) {
            logCompilationError(ctx.toLineColumn(), "Continue statement not within a loop")
        }
    }
}

@ParserRule("InvalidOperandsInBinaryOperator", "Checks for invalid operands in binary operators")
class InvalidOperandsInBinaryOperator(errorListener: SlangParserErrorListener) : CompilationRule(errorListener) {

    private fun isInvalidOperand(ctx: ParserRuleContext): Boolean {
        return ctx is SlangParser.FunAnonymousPureExprContext || ctx is SlangParser.FunAnonymousImpureExprContext
                || ctx is SlangParser.ReadInputExprContext
    }

    private fun validateOperands(operand1: ParserRuleContext, operand2: ParserRuleContext, type: String) {
        if (isInvalidOperand(operand1)) {
            logCompilationError(operand1.toLineColumn(), "${operand1.text} cannot be used as operand in $type operation")
        }
        if (isInvalidOperand(operand2)) {
            logCompilationError(operand2.toLineColumn(), "${operand2.text} cannot be used as operand in $type operation")
        }
    }

    override fun enterArithmeticExpr(ctx: SlangParser.ArithmeticExprContext) {
        if (ctx.expr().size == 2) {
            validateOperands(ctx.expr(0), ctx.expr(1), "arithmetic")
        }
    }

    override fun enterComparisonExpr(ctx: SlangParser.ComparisonExprContext) {
        if (ctx.expr().size == 2) {
            validateOperands(ctx.expr(0), ctx.expr(1), "comparison")
        }
    }

    override fun enterBooleanExpr(ctx: SlangParser.BooleanExprContext) {
        if (ctx.expr().size == 2) {
            validateOperands(ctx.expr(0), ctx.expr(1), "boolean")
        }
    }
}

@ParserRule("ValidateScope", "Checks for duplicate variables in the same scope")
class ValidateScope(errorListener: SlangParserErrorListener) : CompilationRule(errorListener) {
    // structs are defined on the global scope. Nested structs are not supported.
    private val compilationUnitStack = ArrayDeque<SlangParser.CompilationUnitContext>()
    private var variableScopes = ArrayDeque<MutableSet<String>>()

    override fun enterStructStmt(ctx: SlangParser.StructStmtContext) {
        if (SymbolTable.structDeclarations.containsValue(ctx)) {
            logCompilationError(ctx.toLineColumn(), "Struct ${ctx.ID().text} already declared")
        } else {
            val compilationUnit = compilationUnitStack.last()
            SymbolTable.structDeclarations[compilationUnit] = ctx
        }
    }

    override fun enterLetExpr(ctx: SlangParser.LetExprContext) {
        val currentScope = variableScopes.last()
        val id = ctx.ID().text
        if (currentScope.contains(id)) {
            logCompilationError(ctx.toLineColumn(), "Variable $id already declared in this scope")
        } else {
            currentScope.add(id)
        }
    }

    override fun enterBlockStmt(ctx: SlangParser.BlockStmtContext) {
        variableScopes.add(mutableSetOf())
    }

    override fun exitBlockStmt(ctx: SlangParser.BlockStmtContext?) {
        variableScopes.removeLast()
    }

    override fun enterCompilationUnit(ctx: SlangParser.CompilationUnitContext) {
        variableScopes.add(mutableSetOf())
        compilationUnitStack.add(ctx)
    }

    override fun exitCompilationUnit(ctx: SlangParser.CompilationUnitContext) {
        variableScopes.removeLast()
        compilationUnitStack.removeLast()
    }
}

