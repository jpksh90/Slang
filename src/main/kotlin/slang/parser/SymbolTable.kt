package slang.parser

object SymbolTable {
    val structDeclarations = mutableMapOf<SlangParser.CompilationUnitContext, SlangParser.StructStmtContext>()
    val globalVariables = mutableMapOf<SlangParser.CompilationUnitContext, String>()
}