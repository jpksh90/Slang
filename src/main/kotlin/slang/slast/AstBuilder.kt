package slang.slast

import SlangBaseVisitor
import SlangLexer
import SlangParser
import org.antlr.v4.runtime.ANTLRInputStream
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.ParserRuleContext


class SlastBuilder(ctx: SlangParser.CompilationUnitContext) {
    val compilationUnit: CompilationUnit

    init {
        val IRBuilder = IRBuilder()
        compilationUnit = IRBuilder.visit(ctx) as CompilationUnit
    }

    class IRBuilder : SlangBaseVisitor<SlastNode>() {

        private fun createSourceCodeInfo(ctx: ParserRuleContext): SourceCodeInfo {
            return SourceCodeInfo(
                ctx.start.line,
                ctx.stop.line,
                ctx.start.charPositionInLine,
                ctx.stop.charPositionInLine
            )
        }

        override fun visitLetExpr(ctx: SlangParser.LetExprContext): SlastNode {
            val expr = LetStmt(ctx.ID().text, visit(ctx.expr()) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitAssignExpr(ctx: SlangParser.AssignExprContext): SlastNode {
            val expr = AssignStmt(visit(ctx.lhs()) as Expr, visit(ctx.expr()) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitFunPureStmt(ctx: SlangParser.FunPureStmtContext): SlastNode {
            val params = ctx.paramList()?.ID()?.map { it.text } ?: emptyList()
            val id = ctx.ID()
            val body = BlockStmt(listOf(ReturnStmt(visit(ctx.expr()) as Expr)))
            val expr = Function(id.text, params, body)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitFunImpureStmt(ctx: SlangParser.FunImpureStmtContext): SlastNode {
            val params = ctx.paramList()?.ID()?.map { it.text } ?: emptyList()
            val body = ctx.stmt().map { visit(it) as Stmt }
            val expr = Function(ctx.ID().text, params, BlockStmt(body))
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }


        override fun visitFunAnonymousPureExpr(ctx: SlangParser.FunAnonymousPureExprContext): SlastNode {
            val params = ctx.paramList()?.ID()?.map { it.text } ?: emptyList()
            val body = BlockStmt(listOf(ReturnStmt(visit(ctx.expr()) as Expr)))
            val expr = InlinedFunction(params, body)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitFunAnonymousImpureExpr(ctx: SlangParser.FunAnonymousImpureExprContext): SlastNode {
            val params = ctx.paramList()?.ID()?.map { it.text } ?: emptyList()
            val body = ctx.stmt().map { visit(it) as Stmt }
            val expr = InlinedFunction(params, BlockStmt(body))
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitWhileStmt(ctx: SlangParser.WhileStmtContext): SlastNode {
            val condition = visit(ctx.expr()) as Expr
            val body = BlockStmt(ctx.blockStmt().stmt().map { visit(it) as Stmt })
            val expr = WhileStmt(condition, body)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitPrintStmt(ctx: SlangParser.PrintStmtContext): SlastNode {
            val args = ctx.argList()?.expr()?.map { visit(it) as Expr } ?: emptyList()
            val expr = PrintStmt(args)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitIfThenElseStmt(ctx: SlangParser.IfThenElseStmtContext): SlastNode {
            val condition = visit(ctx.expr()) as Expr
            val thenBody = BlockStmt(ctx.blockStmt(0).stmt().map { visit(it) as Stmt })
            val elseBody = BlockStmt(ctx.blockStmt(1).stmt().map { visit(it) as Stmt })
            val expr = IfStmt(condition, thenBody, elseBody)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitExprStmt(ctx: SlangParser.ExprStmtContext): SlastNode {
            val expr = ExprStmt(visit(ctx.expr()) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitReturnStmt(ctx: SlangParser.ReturnStmtContext): SlastNode {
            val expr = ReturnStmt(visit(ctx.expr()) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitIntExpr(ctx: SlangParser.IntExprContext): SlastNode {
            val expr = NumberLiteral(ctx.NUMBER().text.toDouble())
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitBoolExpr(ctx: SlangParser.BoolExprContext): SlastNode {
            val expr = BoolLiteral(ctx.BOOL().text.toBoolean())
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitVarExpr(ctx: SlangParser.VarExprContext): SlastNode {
            val expr = VarExpr(ctx.ID().text)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitReadInputExpr(ctx: SlangParser.ReadInputExprContext): SlastNode {
            val expr = ReadInputExpr
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitFuncCallExpr(ctx: SlangParser.FuncCallExprContext): SlastNode {
            val args = ctx.argList()?.expr()?.map { visit(it) as Expr } ?: emptyList()
            val target = visit(ctx.expr()) as Expr
            val expr = if (target is VarExpr) {
                NamedFunctionCall(target.name, args)
            } else {
                ExpressionFunctionCall(target, args)
            }
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitArithmeticExpr(ctx: SlangParser.ArithmeticExprContext): SlastNode {
            val expr =
                BinaryExpr(visit(ctx.expr(0)) as Expr, Operator.fromValue(ctx.op.text), visit(ctx.expr(1)) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitComparisonExpr(ctx: SlangParser.ComparisonExprContext): SlastNode {
            val expr =
                BinaryExpr(visit(ctx.expr(0)) as Expr, Operator.fromValue(ctx.op.text), visit(ctx.expr(1)) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitBooleanExpr(ctx: SlangParser.BooleanExprContext): SlastNode {
            val expr =
                BinaryExpr(visit(ctx.expr(0)) as Expr, Operator.fromValue(ctx.op.text), visit(ctx.expr(1)) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitIfExpr(ctx: SlangParser.IfExprContext): SlastNode {
            val expr = IfExpr(
                visit(ctx.expr(0)) as Expr,
                visit(ctx.expr(1)) as Expr,
                visit(ctx.expr(2)) as Expr
            )
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitParenExpr(ctx: SlangParser.ParenExprContext): SlastNode {
            val expr = ParenExpr(visit(ctx.expr()) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitForStmt(ctx: SlangParser.ForStmtContext): SlastNode {
            val initialization = AssignStmt(visit(ctx.ID()) as Expr, visit(ctx.expr(0)) as Expr)
            val condition = visit(ctx.expr(1)) as Expr
            val update = visit(ctx.expr(2)) as Stmt
            val body = ctx.blockStmt().stmt().map { visit(it) as Stmt }
            val expr = BlockStmt(listOf(initialization, WhileStmt(condition, BlockStmt(body + update))))
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitCompilationUnit(ctx: SlangParser.CompilationUnitContext): SlastNode {
            val expr = if (ctx.stmt() == null) {
                CompilationUnit(emptyList())
            } else {
                val statements = ctx.stmt().map { visit(it) as Stmt }
                CompilationUnit(statements)
            }
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitNoneValue(ctx: SlangParser.NoneValueContext): SlastNode {
            val expr = NoneValue
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitRecordExpr(ctx: SlangParser.RecordExprContext): SlastNode {
            val recordIds = ctx.recordElems().ID()
            val recordExprs = ctx.recordElems().expr()
            val recordElementPairs = mutableListOf<Pair<String, Expr>>()
            for (i in 0 until recordIds.size) {
                recordElementPairs.addLast(Pair(recordIds[i].text, visit(recordExprs[i]) as Expr))
            }
            val expr = Record(recordElementPairs)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitStringExpr(ctx: SlangParser.StringExprContext): SlastNode {
            val expr = StringLiteral(ctx.STRING().text)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitRefExpr(ctx: SlangParser.RefExprContext): SlastNode {
            val expr = RefExpr(visit(ctx.expr()) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitDerefExpr(ctx: SlangParser.DerefExprContext): SlastNode {
            val expr = DerefExpr(visit(ctx.deref()) as Expr)
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitDeref(ctx: SlangParser.DerefContext): SlastNode {
            val expr = visit(ctx.expr())
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitLhs(ctx: SlangParser.LhsContext): SlastNode {
            val expr = if (ctx.deref() != null) {
                DerefExpr(visit(ctx.deref()) as Expr)
            } else if (ctx.fieldAccess() != null) {
                visit(ctx.fieldAccess())
            } else {
                VarExpr(ctx.ID().text)
            }
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitPrimaryExprWrapper(ctx: SlangParser.PrimaryExprWrapperContext): SlastNode {
            val expr = visit(ctx.primaryExpr())
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitFieldAccess(ctx: SlangParser.FieldAccessContext): SlastNode {
            val expr = FieldAccess(visit(ctx.expr()) as Expr, VarExpr(ctx.ID().text))
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitFieldAccessExpr(ctx: SlangParser.FieldAccessExprContext): SlastNode {
            val expr = FieldAccess(visit(ctx.expr()) as Expr, VarExpr(ctx.ID().text))
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitDoWhileStmt(ctx: SlangParser.DoWhileStmtContext): SlastNode {
            val body = visit(ctx.blockStmt()) as Stmt
            val condition = visit(ctx.expr())
            val expr = BlockStmt(listOf(body, WhileStmt(condition as Expr, body as BlockStmt)))
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }

        override fun visitBlockStmt(ctx: SlangParser.BlockStmtContext): SlastNode {
            val expr = BlockStmt(ctx.stmt().map { visit(it) as Stmt })
            expr.sourceCodeInfo = createSourceCodeInfo(ctx)
            return expr
        }
    }
}

fun main() {
    val x = "fun apply_n(f, n, x) => if (n == 0) then x else f(apply_n(f, n - 1, f(x)));".trimMargin()

    fun parseProgram(input: String): SlastNode {
        val parser = SlangParser(CommonTokenStream(SlangLexer(ANTLRInputStream(input))))

        parser.removeErrorListeners()
        val errorListener = CustomErrorListener()
        parser.addErrorListener(errorListener)

        val parseTree = parser.compilationUnit()

        val IRBuilder = SlastBuilder.IRBuilder()
        val ast = IRBuilder.visit(parseTree) as CompilationUnit
        return ast
    }

    print(parseProgram(x))

}