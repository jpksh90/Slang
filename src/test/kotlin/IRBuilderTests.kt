import org.approvaltests.Approvals
import slang.parser.Parser
import slang.slast.SlastBuilder
import slang.slast.prettyPrint
import java.io.File
import java.net.URL
import kotlin.test.Test

class IRBuilderTests {

    fun loadTestProgram(name: String): URL? {
        return IRBuilderTests::class.java.getResource("/$name")
    }

    fun buildAst(testCase: URL): String {
        val file = File(testCase.toURI())
        val parser = Parser(file)
        val result = parser.parse()
        if (result) {
            val irBuilder = SlastBuilder(parser.compilationUnit).compilationUnit
            return irBuilder.prettyPrint()
        }
        return ""
    }

    @Test
    fun testCase1() {
        val program = loadTestProgram("exponent.slang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase2() {
        val program = loadTestProgram("facl.slang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase3() {
        val program = loadTestProgram("gcd.slang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase4() {
        val program = loadTestProgram("minmax.slang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase5() {
        val program = loadTestProgram("disallowed.slangang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase6() {
        val program = loadTestProgram("sum_prod.slang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase7() {
        val program = loadTestProgram("do-while.slangang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }

    @Test
    fun testCase8() {
        val program = loadTestProgram("misc.slang")
        if (program != null) {
            val ast = buildAst(program)
            Approvals.verify(ast)
        }
    }
}