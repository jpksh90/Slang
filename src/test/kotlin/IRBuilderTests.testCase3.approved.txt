 fun mod(a, b) {
  {
     return if (a < b) then a else mod(a - b, b);
  }
}
 fun gcd(a, b) {
  {
     return if (b == 0.0) then a else gcd(b, mod(a, b));
  }
}